sdk/identity/azure-identity/src/azure_cli_credential.cpp

```diff
@@ -0,0 +1,654 @@
+
+using Azure::DateTime;
+using Azure::Core::Context;
+using Azure::Core::_internal::Environment;
+using Azure::Core::Credentials::AccessToken;
+using Azure::Core::Credentials::AuthenticationException;
+using Azure::Core::Credentials::TokenCredentialOptions;
+using Azure::Core::Credentials::TokenRequestContext;
+using Azure::Identity::AzureCliCredentialOptions;
+using Azure::Identity::_detail::TokenCredentialImpl;
+
+namespace {
+void ThrowIfNotSafeCmdLineInput(std::string const& input, std::string const& description)
+{
+  for (auto const c : input)
+  {
+    switch (c)
+    {
+      case ':':
+      case '/':
+      case '.':
+      case '-':
+      case '_':
+      case ' ':
+        break;
+
+      default:
+        if (!std::isalnum(c))
+        {
+          throw AuthenticationException(
+              "AzureCliCredential: Unsafe command line input found in " + description + ": "
+              + input);
+        }
+    }
+  }
+}
+} // namespace
+
+AzureCliCredential::AzureCliCredential(
+    std::string tenantId,
+    DateTime::duration cliProcessTimeout,
+    Core::Credentials::TokenCredentialOptions const& options)
+    : m_tenantId(std::move(tenantId)), m_cliProcessTimeout(std::move(cliProcessTimeout))
+{
+  static_cast<void>(options);
+  ThrowIfNotSafeCmdLineInput(m_tenantId, "TenantID");
+}
+
+AzureCliCredential::AzureCliCredential(AzureCliCredentialOptions const& options)
+    : AzureCliCredential(options.TenantId, options.CliProcessTimeout, options)
+{
+}
+
+AzureCliCredential::AzureCliCredential(TokenCredentialOptions const& options)
+    : AzureCliCredential(
+        AzureCliCredentialOptions{}.TenantId,
+        AzureCliCredentialOptions{}.CliProcessTimeout,
+        options)
+{
+}
+
+std::string AzureCliCredential::GetAzCommand(
+    std::string const& resource,
+    std::string const& tenantId) const
+{
+  ThrowIfNotSafeCmdLineInput(resource, "Resource");
+  std::string command = "az account get-access-token --output json --resource \"" + resource + "\"";
+
+  if (!tenantId.empty())
+  {
+    command += " --tenant \"" + tenantId + "\"";
+  }
+
+  return command;
+}
+
+namespace {
+std::string RunShellCommand(
+    std::string const& command,
+    DateTime::duration timeout,
+    Context const& context);
+}
+
+AccessToken AzureCliCredential::GetToken(
+    TokenRequestContext const& tokenRequestContext,
+    Context const& context) const
+{
+  try
+  {
+    auto const azCliResult = RunShellCommand(
+        GetAzCommand(
+            TokenCredentialImpl::FormatScopes(tokenRequestContext.Scopes, true, false), m_tenantId),
+        m_cliProcessTimeout,
+        context);
+
+    try
+    {
+      return TokenCredentialImpl::ParseToken(azCliResult, "accessToken", "expiresIn", "expiresOn");
+    }
+    catch (std::exception const&)
+    {
+      // Throw the az command output (error message)
+      // limited to 250 characters (250 has no special meaning).
+      throw std::runtime_error(azCliResult.substr(0, 250));
```

Comment:
Why isn't this an AuthenticationException?

Reply:
Also, catching an exception and discarding it is usually considered bad form - you should at least include the `exception.what()` in the rethrown exception.

Reply:
`TokenCredentialImpl::ParseToken()` would only throw `std::exception`. And the only reason it would throw, is that it failed to parse the token. If it failed to parse, then the output that was given to it is probably not JSON token, but a string like `ERROR: Please run 'az login' to setup account.`. Then it makes sense to grab that input and put that to the exception message, instead of the `ex.what()` of the exception that was thrown from `ParseToken()`, which usually looks like `"Token JSON object: 'accessToken' property was not found."`.

And we don't throw `AuthenticationException` right here is that because below we wrap all the `std::exception`s as `AuthenticationException` by prepending `AzureCliCredential: ` in the front (line 145). If we want to throw `AuthenticationException` from right here, we would duplicate that prefix prepending logic, and would need to add rethrow block for the try-catch block below. If we leave it as is, optimizer may do it for us. Plus, performance in exception case is not critical, so I optimized for less code complexity.

I.e. if `azCliResult` (the text we got from invoking the `az account`) is `ERROR: Please run 'az login' to setup account.`, then the exception that gets thrown from this function is `AuthenticationException`, with the `what()` == `"AzureCliCredential::GetToken(): ERROR: Please run 'az login' to setup account."`